File: tick-collector.js
const db = require('./db');

const collectors = {};
const lastTicks = {}; // symbol -> { price, volume }
const POLL_INTERVAL_MS = 1000;
const MAX_COLLECTOR_RUN_MS = 2 * 60 * 1000; // auto-stop after 2 minutes

function startCollector(symbol, token, smartApi) {
  const key = symbol.toUpperCase();

  if (collectors[key]) {
    return false;
  }

  if (!smartApi || typeof smartApi.marketData !== 'function') {
    console.error(`[TickCollector] Cannot start collector for ${key}: smartApi not available`);
    return false;
  }

  console.log(`[TickCollector] Starting collector for ${key} (token: ${token}, interval: ${POLL_INTERVAL_MS}ms)`);

  const startedAt = Date.now();

  const intervalId = setInterval(async () => {
    try {
      // Auto-stop after max run time
      if (Date.now() - startedAt > MAX_COLLECTOR_RUN_MS) {
        console.log(`[TickCollector/${key}] Auto-stopping after ${MAX_COLLECTOR_RUN_MS / 1000}s`);
        stopCollector(key);
        return;
      }

      const resp = await smartApi.marketData({
        mode: "FULL",
        exchangeTokens: { NSE: [token] },
        tradingsymbol: `${key}-EQ`
      });

      let tick = null;
      if (resp && resp.data && resp.data.fetched && resp.data.fetched.length > 0) {
        tick = resp.data.fetched[0];
      } else if (resp && resp.fetched && resp.fetched.length > 0) {
        tick = resp.fetched[0];
      }

      if (tick && (tick.ltp != null || tick.close != null)) {
        if (!lastTicks[key]) {
          console.log(`[TickCollector/${key}] Raw tick fields:`, Object.keys(tick).join(', '));
          console.log(`[TickCollector/${key}] Sample tick:`, JSON.stringify(tick));
        }

        const now = new Date();
        const fallbackStr = db.formatTimestamp(now);
        const rawTime = tick.exchangeTime || tick.lastTradeTime || tick.exchTradeTime || fallbackStr;
        const price = tick.ltp != null ? tick.ltp : tick.close;
        const volume = tick.lastTradedQuantity || tick.volume || 0;
        if (!tick.lastTradedQuantity && !tick.volume && !lastTicks[key]) {
          console.log(`[TickCollector/${key}] WARNING: getLtpData endpoint returns no volume field. Volume will be 0.`);
        }

        const last = lastTicks[key];
        if (last && last.price === price && last.volume === volume) {
          return;
        }
        lastTicks[key] = { price, volume };

        await db.insertTick(key, price, volume, rawTime);
      }
    } catch (err) {
      // suppress log spam on transient network errors
    }
  }, POLL_INTERVAL_MS);

  collectors[key] = {
    symbol: key,
    token,
    intervalId,
    startedAt,
    pollCount: 0
  };

  return true;
}

function stopCollector(symbol) {
  const key = symbol.toUpperCase();
  const collector = collectors[key];
  if (!collector) return false;

  clearInterval(collector.intervalId);
  delete collectors[key];
  console.log(`[TickCollector] Stopped collector for ${key}`);
  return true;
}

function isCollectorRunning(symbol) {
  const key = symbol.toUpperCase();
  return !!collectors[key];
}

function getActiveCollectors() {
  return Object.keys(collectors);
}

async function getCurrentBucketOHLC(symbol) {
  const key = symbol.toUpperCase();
  const now = new Date();
  const bucketStart = db.getFiveMinuteBucket(now);
  const bucketEnd = new Date(bucketStart);
  bucketEnd.setMinutes(bucketStart.getMinutes() + 5);

  const fromStr = db.formatTimestamp(bucketStart);
  const toStr = db.formatTimestamp(bucketEnd);

  const rows = await db.getTicksInRange(key, fromStr, toStr);
  return db.aggregateOHLCFromTicks(rows);
}

function getBucketBoundaries(date) {
  const now = date || new Date();
  const bucketStart = db.getFiveMinuteBucket(now);
  const bucketEnd = new Date(bucketStart);
  bucketEnd.setMinutes(bucketStart.getMinutes() + 5);
  return {
    start: bucketStart,
    end: bucketEnd,
    startStr: db.formatTimestamp(bucketStart),
    endStr: db.formatTimestamp(bucketEnd)
  };
}

async function getLiveCandleForBucket(symbol, bucketStartDate) {
  const key = symbol.toUpperCase();
  const bucketStart = db.getFiveMinuteBucket(bucketStartDate || new Date());
  const bucketEnd = new Date(bucketStart);
  bucketEnd.setMinutes(bucketStart.getMinutes() + 5);

  const fromStr = db.formatTimestamp(bucketStart);
  const toStr = db.formatTimestamp(bucketEnd);

  const rows = await db.getTicksInRange(key, fromStr, toStr);
  if (!rows || rows.length === 0) return null;

  const ohlc = db.aggregateOHLCFromTicks(rows);
  return {
    datetime: fromStr,
    open: ohlc.open,
    high: ohlc.high,
    low: ohlc.low,
    close: ohlc.close,
    volume: ohlc.volume,
    tickCount: rows.length
  };
}

function stopAllCollectors() {
  Object.keys(collectors).forEach(key => stopCollector(key));
}

module.exports = {
  startCollector,
  stopCollector,
  isCollectorRunning,
  getActiveCollectors,
  getCurrentBucketOHLC,
  getBucketBoundaries,
  getLiveCandleForBucket,
  stopAllCollectors
};

File: fastapi_app/tick_collector.py
import asyncio
import time
import json
import os
from datetime import datetime, timedelta
import angelone_client
from database import insert_tick, format_timestamp, get_five_minute_bucket

collectors = {}
last_ticks = {}
POLL_INTERVAL_MS = 1000
MAX_COLLECTOR_RUN_MS = 2 * 60 * 1000

async def collector_task(symbol: str, token: str):
    key = symbol.upper()
    started_at = time.time() * 1000
    
    while True:
        if key not in collectors:
            break
            
        if (time.time() * 1000) - started_at > MAX_COLLECTOR_RUN_MS:
            print(f"[TickCollector/{key}] Auto-stopping after {MAX_COLLECTOR_RUN_MS / 1000}s")
            stop_collector(key)
            break
            
        try:
            # We call fetchMarketData manually via aiohttp here or through angelone_client
            # Wait, angelone_client doesn't have fetchMarketData exported. I need to use the endpoint.
            await angelone_client.ensure_logged_in()
            import aiohttp
            
            headers = {
                "Content-Type": "application/json",
                "Accept": "application/json",
                "X-UserType": "USER",
                "X-SourceID": "WEB",
                "X-ClientLocalIP": "127.0.0.1",
                "X-ClientPublicIP": "127.0.0.1",
                "X-MACAddress": "01-01-01-01-01-01",
                "X-PrivateKey": angelone_client.ANGELONE_CLIENT_SECRET,
                "Authorization": f"Bearer {angelone_client.jwt_token}",
            }
            payload = {
                "exchange": "NSE",
                "tradingsymbol": f"{key}-EQ",
                "symboltoken": token
            }
            async with aiohttp.ClientSession() as session:
                async with session.post("https://apiconnect.angelbroking.com/rest/secure/angelbroking/order/v1/getLtpData", json=payload, headers=headers) as resp:
                    resp_data = await resp.json()
            
            tick = None
            if resp_data and resp_data.get("data"):
                tick = resp_data["data"]
                
            if tick and (tick.get("ltp") is not None or tick.get("close") is not None):
                now = datetime.now()
                raw_time = tick.get("exchangeTime", tick.get("lastTradeTime", format_timestamp(now)))
                price = float(tick.get("ltp", tick.get("close", 0)))
                open_p = float(tick.get("open", price))
                high = float(tick.get("high", price))
                low = float(tick.get("low", price))
                close = float(tick.get("close", price))
                volume = int(tick.get("lastTradedQuantity", tick.get("volume", 0)))
                
                last = last_ticks.get(key)
                if last and last["close"] == close and last["volume"] == volume and last["open"] == open_p and last["high"] == high and last["low"] == low:
                    pass
                else:
                    last_ticks[key] = {"open": open_p, "high": high, "low": low, "close": close, "volume": volume}
                    await insert_tick(key, open_p, high, low, close, volume, raw_time)
                    msg = f"[TickCollector] Received Live Tick -> {key}: Open={open_p}, High={high}, Low={low}, Close={close}, Vol={volume}, Time={raw_time}"
                    print(msg)
                    with open(os.path.join(os.path.dirname(__file__), 'tick_collector.log'), 'a') as f:
                        f.write(f"[{datetime.now().isoformat()}] {msg}\n")
                        
        except Exception as e:
            pass # suppress transient network errors
            
        await asyncio.sleep(POLL_INTERVAL_MS / 1000.0)

def start_collector(symbol: str, token: str):
    key = symbol.upper()
    if key in collectors:
        return False
        
    print(f"[TickCollector] Starting collector for {key} (token: {token}, interval: {POLL_INTERVAL_MS}ms)")
    task = asyncio.create_task(collector_task(key, token))
    collectors[key] = {
        "symbol": key,
        "token": token,
        "task": task,
        "startedAt": time.time() * 1000
    }
    return True

def stop_collector(symbol: str):
    key = symbol.upper()
    if key in collectors:
        collectors[key]["task"].cancel()
        del collectors[key]
        print(f"[TickCollector] Stopped collector for {key}")
        return True
    return False

def is_collector_running(symbol: str):
    return symbol.upper() in collectors

def get_active_collectors():
    return list(collectors.keys())

def stop_all_collectors():
    keys = list(collectors.keys())
    for key in keys:
        stop_collector(key)
